src/components/FallenHeroes.tsx

```diff
@@ -20,23 +34,27 @@ function FallenHeroes() {
   function addFallenHero(e: React.FormEvent<HTMLFormElement>) {
     e.preventDefault();
     //add hero to the list if it isnt there yet
-    if (!campaign.restOfAliveHeroes.includes(fallenHero)) {
+    if (!names.includes(fallenHero)) {
+      throw Error("Sorry, your hero is an unknown!");
+    }
+
+    if (names.includes(fallenHero) && !isAlive) {
       throw Error("Sorry, your hero is not among the living!");
     }
-    if (campaign.restOfAliveHeroes.includes(fallenHero)) {
-      const updatedFallen = [...campaign.fallenHeroes, fallenHero];
-      const updatedRestOfAlive = campaign.restOfAliveHeroes.filter(
-        (hero) => hero !== fallenHero
-      );
-      setCampaign({
-        ...campaign,
-        fallenHeroes: updatedFallen,
-        restOfAliveHeroes: updatedRestOfAlive,
+
+    if (names.includes(fallenHero) && isAlive) {
+      campaign.allHeroes.map(function (hero) {
+        if (hero.name === fallenHero) {
+          hero.alive = false;
+          setCampaign(campaign);
+          const fallen = filterFallen();
+          setFallenHeroes(fallen);
```

Comment:
Once you address the other issues, you can remove these lines and just use the `fallen` variable. It is a derived value, no need to keep it in state.l